src/smallOrLargeNodes/content.md

```diff
@@ -0,0 +1,293 @@
+
+If you use large nodes, then you have a large scaling increment, which makes scaling more clunky.
+
+For example, if you only have 2 nodes, then adding an additional node means increasing the capacity of the cluster by 50%.
+
+This might be much more than you actually need, and you pay for unused resources.
+
+So, if you plan to automatically scale your cluster, then smaller nodes will enable a more fluid and cost-efficient scaling behaviour.
+
+_Having discussed the pros and cons of few large nodes, let's turn to the scenario of many small nodes._
+
+## Many small nodes
+
+The second approach is to achieve your desired cluster capacity by assembling many small nodes, instead of few large nodes.
+
+_What are the pros and cons of this approach?_
+
+### Pros
+
+The pros of using many small nodes correspond mainly to the cons of using few large nodes.
+
+#### Reduced blast radius
+
+If you have more nodes, you naturally have fewer pods on each node.
+
+For example, if you have 100 pods and 10 nodes, then each node contains on average only 10 pods.
+
+Thus, if one of the nodes fails, the impact is limited to a smaller proportion of your total workload.
+
+Chances are that only some of your apps are affected, and potentially only a small number of replicas so that the apps as a whole stay up.
+
+Furthermore, there are most likely enough spare resources on the remaining nodes to accommodate the workload of the failed node, so that Kubernetes can reschedule all the pods and your apps return to a fully functional state relatively quickly.
+
+#### Allows high replication
+
+If you have replicated high-availability apps, and enough available nodes, the Kubernetes scheduler can assign each replica to a different node.
+
+> You can influence scheduler's the placement of pods with [node and pod affinities/anti-affinities](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) and with [taints and tolerations](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/).
+
+This means that if a node fails, there is at most one replica affected and your app stays available.
+
+_Having seen the pros of using many small nodes, what are the cons?_
+
+### Cons
+
+#### Large number of nodes
+
+If you use smaller nodes, you naturally need more of them to achieve a given cluster capacity.
+
+_But large numbers of nodes can be a challenge for the Kubernetes control plane._
+
+For example, every node needs to be able to communicate with every other node, which makes the number of possible communication paths grow exponentially with the number of nodes — all of which has to be managed by the control plane.
+
+The node controller in the Kubernetes controller manager regularly has to iterate through all the nodes in the cluster and run health checks — more nodes mean thus more load for the node controller.
```

Comment:
I had this from "Kubernetes in Action": _Among other things, a Node controller keeps the list of Node objects in sync with the actual list of machines running in the cluster. It also monitors each node’s health and evicts pods from unreachable nodes._

I think the kubelets also report some things back to the Node objects in etcd. But the node controller probes the nodes certainly too. Because if the whole node disappears, the kubelet disappears too, so then only the node controller can detect that the node is down.